### PR TITLE
Add ParticleSystem CellIndex Overlife

### PR DESCRIPTION
**Pull request by Moriyer**
Add a new field spriteCellIndexOverlife to basePraticleSystem, it allows to use hermite curve to convert cellIndex overlife.
It has some changes with Inspector to editor this field in Inspector.
It has some changes with Animation Curve Editor and Contex to add a way to allow the last key's frame/value can't be changed.

### CODE REVIEW

File: inspector/src/components/actionTabs/tabs/propertyGrids/particleSystems/HermiteCurveComponent.tsx [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+        //             <ButtonLineComponent
+        //                 label={"Use " + this.props.label}
+        //                 onClick={() => {
+        //                     this.props.onCreateRequired();
+        //                     this.forceUpdate();
+        //                 }}
+        //             />
+        //         )}
+        //     </div>
+        // );
```
Comment (RaananW) [inline inspector/src/components/actionTabs/tabs/propertyGrids/particleSystems/HermiteCurveComponent.tsx:240]: Is that still needed?

File: src/Misc/HermiteCurve.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            return this.keys[0].value;
+        }
+        if (process <= 0) {
+            return this.keys[0].value;
+        }
+        if (process >= 1) {
+            return this.keys[this.keys.length - 1].value;
+        }
+        for (var key = 1; key < this.keys.length; key++) {
+            var endKey = this.keys[key];
```
Comment (RaananW) [inline src/Misc/HermiteCurve.ts:52]: nit! :-)
```suggestion
        for (let key = 1; key < this.keys.length; key++) {
            const endKey = this.keys[key];
```

File: inspector/src/components/actionTabs/tabs/propertyGrids/particleSystems/HermiteCurveComponent.tsx [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+            const animKey = this._animation!.getKeys()[keyId];
+            key.value = animKey.value;
+            key.frame = animKey.frame / frameScale;
+            key.inTangent = animKey.inTangent! * frameScale;
+            key.outTangent = animKey.outTangent! * frameScale;
+        }
+        else {
+            const newKeys: IAnimationKey[] = [];
+            animKeys.forEach(key => {
+                newKeys.push({
```
Comment (RaananW) [inline inspector/src/components/actionTabs/tabs/propertyGrids/particleSystems/HermiteCurveComponent.tsx:106]: nit - can be using `.map` instead something like:

```javascript 
const newKeys = animKeys.map( key => { /// animation key here})

**Comment by Moriyer** [comment]
I will change it soon, I have a question that, I add HermiteCurve in misc, how I get it ts reference and new it in&nbsp;HermiteCurveComponent.tsx




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "BabylonJS/Babylon.js"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年1月5日(星期三) 晚上8:27
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [BabylonJS/Babylon.js] Add ParticleSystem CellIndex Overlife (PR #11711)





 
@RaananW commented on this pull request.
 
 
In inspector/src/components/actionTabs/tabs/propertyGrids/particleSystems/HermiteCurveComponent.tsx:
 &gt; +        // let gradients = this.props.gradients as Nullable<Array<IValueGradient&gt;&gt;; + +        // return ( +        //     <div&gt; +        //         {gradients &amp;&amp; gradients.length &gt; 0 &amp;&amp; ( +        //             <div className="gradient-container"&gt; +        //                 <LinkButtonComponent +        //                     label={this.props.label} +        //                     url={this.props.docLink} +        //                     icon={faTrash} +        //                     onIconClick={() =&gt; { +        //                         gradients!.length = 0; +        //                         this.updateAndSync(); +        //                     }} +        //                     buttonLabel="Add new step" +        //                     onClick={() =&gt; this.addNewStep()} +        //                 /&gt; +        //                 {gradients.map((g, i) =&gt; { +        //                     let codeRecorderPropertyName = this.props.codeRecorderPropertyName + `[${i}]`; +        //                     switch (this.props.mode) { +        //                         case GradientGridMode.Factor: +        //                             return ( +        //                                 <FactorGradientStepGridComponent +        //                                     globalState={this.props.globalState} +        //                                     lockObject={this.props.lockObject} +        //                                     onCheckForReOrder={() =&gt; this.checkForReOrder()} +        //                                     onUpdateGradient={() =&gt; this.updateAndSync()} +        //                                     host={this.props.host} +        //                                     codeRecorderPropertyName={codeRecorderPropertyName} +        //                                     key={"step-" + i} +        //                                     lineIndex={i} +        //                                     gradient={g as FactorGradient} +        //                                     onDelete={() =&gt; this.deleteStep(g)} +        //                                 /&gt; +        //                             ); +        //                         case GradientGridMode.Color4: +        //                             return ( +        //                                 <ColorGradientStepGridComponent +        //                                     globalState={this.props.globalState} +        //                                     host={this.props.host} +        //                                     codeRecorderPropertyName={codeRecorderPropertyName} +        //                                     lockObject={this.props.lockObject} +        //                                     isColor3={false} +        //                                     onCheckForReOrder={() =&gt; this.checkForReOrder()} +        //                                     onUpdateGradient={() =&gt; this.updateAndSync()} +        //                                     key={"step-" + i} +        //                                     lineIndex={i} +        //                                     gradient={g as ColorGradient} +        //                                     onDelete={() =&gt; this.deleteStep(g)} +        //                                 /&gt; +        //                             ); +        //                         case GradientGridMode.Color3: +        //                             return ( +        //                                 <ColorGradientStepGridComponent +        //                                     globalState={this.props.globalState} +        //                                     host={this.props.host} +        //                                     codeRecorderPropertyName={codeRecorderPropertyName} +        //                                     lockObject={this.props.lockObject} +        //                                     isColor3={true} +        //                                     onCheckForReOrder={() =&gt; this.checkForReOrder()} +        //                                     onUpdateGradient={() =&gt; this.updateAndSync()} +        //                                     key={"step-" + i} +        //                                     lineIndex={i} +        //                                     gradient={g as Color3Gradient} +        //                                     onDelete={() =&gt; this.deleteStep(g)} +        //                                 /&gt; +        //                             ); +        //                     } +        //                 })} +        //             </div&gt; +        //         )} +        //         {(!gradients || gradients.length === 0) &amp;&amp; ( +        //             <ButtonLineComponent +        //                 label={"Use " + this.props.label} +        //                 onClick={() =&gt; { +        //                     this.props.onCreateRequired(); +        //                     this.forceUpdate(); +        //                 }} +        //             /&gt; +        //         )} +        //     </div&gt; +        // );  
Is that still needed?
 
 
In src/Misc/HermiteCurve.ts:
 &gt; +        for (var key = 1; key < this.keys.length; key++) { +            var endKey = this.keys[key];  
nit! :-)
 ⬇️ Suggested change -        for (var key = 1; key < this.keys.length; key++) { -            var endKey = this.keys[key]; +        for (let key = 1; key < this.keys.length; key++) { +            const endKey = this.keys[key];  
 
In inspector/src/components/actionTabs/tabs/propertyGrids/particleSystems/HermiteCurveComponent.tsx:
 &gt; +            const newKeys: IAnimationKey[] = []; +            animKeys.forEach(key =&gt; { +                newKeys.push({  
nit - can be using .map instead something like:
 const newKeys = animKeys.map( key =&gt; { /// animation key here})
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by sebavan** [comment]
> I will change it soon, I have a question that, I add HermiteCurve in misc, how I get it ts reference and new it in&nbsp;HermiteCurveComponent.tsx

I d say this would be a different PR for the curve editor currently, let s only check the curve system for particles ?

But this question will be for @deltakosh :-)

**Comment by sebavan** [comment]
I am flagging as draft for now until we follow up with @PatrickRyanMS in https://forum.babylonjs.com/t/add-gradient-curve-for-particlesystem-spirtesheet-cellindex-over-life/26367

**Comment by deltakosh** [comment]
I'm not super fan of this PR to be honest. First it does not work with GPU particles. Then we are doing too much in one PR (we should first focus on the feature, see if it makes sense and then update the tools)

Furthermore I'm not sold on the need of a specific case for cellIndex. We already have other parameters changing over time and I don't see why we need a special case for cellIndex.

**Comment by sebavan** [comment]
Let me actually close the PR for now and take back the discussion to the forum to help you create the feature.

**Pull request closed without merging**